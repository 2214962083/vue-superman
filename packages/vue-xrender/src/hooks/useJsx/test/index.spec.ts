import {defineComponent, h, nextTick, ref} from 'vue-demi'
import {useJsx} from '../'
import {mount, getComponent} from '@test/mount'

const createComponent = () =>
  defineComponent({
    setup() {
      const count = ref(0)
      const countIncrement = () => count.value++

      const SubTitle = useJsx(h('h3', {class: 'sub-title'}, [`subTitle: count first frame value: ${count.value}`]))

      const Title = useJsx(props =>
        h('div', {class: 'title'}, [h('h1', [`title: count realtime value: ${count.value}`]), props.children])
      )

      const Card2 = useJsx('MyCard', props => {
        return h('div', {class: 'card'}, [
          h('h1', ['Card component, create in setup useJsx, use in template']),
          h(Title, [h(SubTitle)]),
          props.children
        ])
      })

      return {
        count,
        countIncrement,
        SubTitle,
        Title,
        Card2
      }
    },
    render() {
      const Card = getComponent('my-card')
      return h(Card, [h('div', {class: 'card-children'}, ['Card components children from template'])])
    }
  })

describe('useJsx', () => {
  test('should return vue component', async () => {
    const vm = mount(createComponent())

    const subTitle = vm.find('.sub-title')
    const title = vm.find('.title')
    const card = vm.find('.card')
    vm.countIncrement()

    await nextTick()

    expect(subTitle.text()).toBe('subTitle: count first frame value: 0')

    expect(title.find('h1').text()).toBe('title: count realtime value: 1')
    expect(title.find('.sub-title').text()).toBe('subTitle: count first frame value: 0')

    expect(card.find('h1').text()).toBe('Card component, create in setup useJsx, use in template')
    expect(card.find('.title').find('h1').text()).toBe('title: count realtime value: 1')
    expect(card.find('.card-children').text()).toBe('Card components children from template')
    vm.toSnap()
  })
})
