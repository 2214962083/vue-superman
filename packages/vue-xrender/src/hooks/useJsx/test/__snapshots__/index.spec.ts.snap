// Vitest Snapshot v1

exports[`useJsx hooks > should auto register jsx as component 1`] = `
<div
  class="card"
>
  <h1>
    Card component, create in setup useJsx, use in template
  </h1>
  <div
    class="title"
  >
    <h1>
      title: count realtime value: 1
    </h1>
    
    <h3
      class="sub-title"
    >
      subTitle: count first frame value: 0
    </h3>
    
  </div>
  
  <div
    class="card-children"
  >
    Card components children from template
  </div>
  
</div>
`;
